Use scap library to implement screensharing on X11

mgsloan opened the pull request:
While `scap` does have support for Wayland and Windows, but haven't seen screensharing work properly there yet.  So for now just adding support for X11 screensharing.

WIP branches for enabling wayland and windows support:

* https://github.com/zed-industries/zed/tree/wayland-screenshare
* https://github.com/zed-industries/zed/tree/windows-screenshare


Release Notes:

- Added support for screensharing on X11 (Linux)

mgsloan committed:
WIP addition of Linux screen sharing support via scap library

Co-authored-by: Conrad <conrad@zed.dev>
Co-authored-by: Mikayla <mikayla@zed.dev>

mgsloan committed:
Update scap deps to not duplicate `windows` and `sysinfo` deps

mgsloan committed:
Cleanup progress

mgsloan committed:
Merge branch 'main' into linux-screenshare

mgsloan committed:
Further cleanup

mgsloan committed:
Enable screen share on windows (staff only for now as untested)

mgsloan committed:
Bump scap dep to remove a debug println

mgsloan committed:
Improve error handling + remove some dbg!

mgsloan committed:
Fix screen capture post refactor

mgsloan committed:
Make scap dep conditional

mgsloan committed:
Misc polish

mgsloan committed:
Improve scap error handling

mgsloan committed:
Remove commented out `publish_local_wav_track`

mgsloan committed:
For now only use scap for X11 screen capture

* Wayland screen capture doesn't work for me yet

* Windows support untested and our scap branch probably doesn't yet build due to changes to error handling

Co-authored-by: Conrad <conrad@zed.dev>

mgsloan committed:
Merge branch 'main' into linux-screenshare

mgsloan committed:
Bump scap fork version to add support for freebsd

mgsloan committed:
Add dbus dev packages to `script/linux`

mgsloan committed:
Clippy

mgsloan committed:
Workaround `Box<dyn ScreenCaptureSource>` not being `Send`

Having `trait ScreenCaptureSource: Send` and `trait ScreenCaptureStream: Send worked for linux compilation, but `MacScreenCaptureSource` does not implement `Send`: https://github.com/zed-industries/zed/actions/runs/14187088998/job/39744280748?pr=27807#step:4:145

An alternative might be to have `unsafe impl Send` for `MacScreenCaptureSource` and `MacScreenCaptureStream`. However, I'm guessing this won't work at all due to Cocoa requiring its use be from the UI thread.

mgsloan committed:
Add pipewire dev packages to `script/linux`

mgsloan committed:
Attempt to fix build on windows and mac

JunkuiZhang committed:
Try fix windows

mgsloan committed:
Update scap with `wayland` feature to allow omitting dbus / pipewire deps

Rationale is that:

* It's been tricky to get scap with wayland support passing CI
* The deps for wayland support aren't needed for this initial version that only supports X11

mgsloan committed:
Remove dbus / pipewire from deps installation, only needed for wayland

mgsloan committed:
Merge branch 'main' into linux-screenshare

mgsloan committed:
Add libxcb deps to script/linux

JunkuiZhang committed:
Fix build on Windows

mgsloan committed:
Only include scap dep on linux if `x11` or `wayland` gpui feature is used

mgsloan committed:
Fix build of `remote-server`

JunkuiZhang commented:
A quick test on Windows, the screen sharing shows nothing but green
![image](https://github.com/user-attachments/assets/6fae4806-8909-40da-865e-1dc963c8aec3)

JunkuiZhang committed:
Fix windwos

JunkuiZhang committed:
Bump scap

mgsloan committed:
Attempt to fix build on mac

mgsloan committed:
Disable screen capture on windows since it doesn't work yet

A revert of this change will be on the `windows-screenshare` branch

mgsloan committed:
Have `livekit_client` enable `x11` and `wayland` features

Without this, `cargo check -p livekit_client` would fail on linux due to `PlatformScreenCaptureFrame` being `()`.

mgsloan committed:
Merge branch 'main' into linux-screenshare

mgsloan committed:
Fix build on windows

mgsloan committed:
Add install of `libxcb-devel` for `dnf` / `yum`

mgsloan committed:
Fix build on windows
